receiver/telemetrygeneratorreceiver/generators.go

```diff
@@ -0,0 +1,69 @@
+// Copyright observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetrygeneratorreceiver //import "github.com/observiq/bindplane-agent/receiver/telemetrygeneratorreceiver"
+
+import (
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+// GeneratorType is the type of generator to use, either "logs", "host_metrics", or "windows_events"
+type GeneratorType = string
+
+const (
+	// GeneratorTypeLogs is the generator type for logs
+	GeneratorTypeLogs GeneratorType = "logs"
+
+	// GeneratorTypeHostMetrics is the generator type for host metrics
+	GeneratorTypeHostMetrics GeneratorType = "host_metrics"
+
+	// GeneratorTypeWindowsEvents is the generator type for windows events
+	GeneratorTypeWindowsEvents GeneratorType = "windows_events"
+)
+
+type generator interface {
+	// SupportsType returns true if the generator supports the given component type, either metrics, logs, or traces.
+	SupportsType(component.Type) bool
+	GenerateMetrics() pmetric.Metrics
+	GenerateLogs() plog.Logs
+	GenerateTraces() ptrace.Traces
+
+	// initialize must called when the generator is created. It is used to prevent the need to
+	// recreate the telemetry every generation cycle
+	initialize()
```

Comment:
The design Corbin and I discussed allows for generators that emit more than one telemetry type. Pushing initialization off until after the constructor avoids building the telemetry for a generator that doesn't produce the type the receiver is using.

Reply:
I actually would prefer the initialize be part of the newX so that we don't need two methods to do the additional setup. I'll resolve my comment for this.

Reply:
Ah I missed the map creation part of [this comment](https://github.com/observIQ/bindplane-agent/pull/1506#discussion_r1507851079). Makes sense to me.